tests/tables.py

```diff
@@ -0,0 +1,339 @@
+from tests.util import run_sql
+from dbt.context import providers
+from unittest.mock import patch
+from contextlib import contextmanager
+
+
+class TableComparison:
+    def __init__(self, adapter, unique_schema, database):
+        self.adapter = adapter
+        self.unique_schema = unique_schema
+        self.default_database = database
+        # We need to get this from somewhere reasonable
+        self.quoting = {'database': True, 'schema': True, 'identifier': True}
+
+    def _assert_tables_equal_sql(self, relation_a, relation_b, columns=None):
+        if columns is None:
+            columns = self.get_relation_columns(relation_a)
+        column_names = [c[0] for c in columns]
+        sql = self.adapter.get_rows_different_sql(relation_a, relation_b, column_names)
```

Comment:
This is a funny method: It ultimately calls `COLUMNS_EQUAL_SQL`, a constant / Python f-string defined on the adapter class [here](https://github.com/dbt-labs/dbt-core/blob/2d0b975b6c2023cde219f0a045709a1fa5c6c840/core/dbt/adapters/base/impl.py#L1197-L1223), and reimplemented on other adapters ([e.g. spark](https://github.com/dbt-labs/dbt-spark/blob/63ebe042419a0feb5939c2eb1c010e7e52e8c802/dbt/adapters/spark/impl.py#L368-L400)). Why not call a macro, as other adapter Python methods do, and use our multiple dispatch pattern?

There are more examples of this further down, starting with `_assert_table_row_counts_equal`

Reply:
All of this code was extracted from test/integration/base.py. Quite a large chunk of that was dealing with the table comparisons, so I just copied, formatted and got it to work. I'm up for whatever better practices have evolved since that code was written, and at the least we would need to override this in adapter repos.